QoL changes

kolt54321 opened the pull request:
So these changes should set zoom levels to conform to JPEG limits, allow bringing in a list of links from a batch CSV with a URL list, create (and maintain) a cache/queue and flag downloaded files so no re-downloading occures, and adds metadata tags from the website art description to the file. It also uses that same set of tags to rework the output name for sorting downloads better.

Let's discuss. Thanks!

kolt54321 committed:
Allows list downloading and tracking

External file that has arguments for importing a list of URL's as well as downloading a list in the queue/cache.
After downloading an image it sets the download flag to true - so if the script breaks in middle it will just pick up from where it left off.

Some measures here are redundent and can be consolidated further. A measure is in place to keep the zoom level within JPEG limits.

Small sleep included so we (hopefully) don't hit 429 errors.

kolt54321 committed:
Add embed metadata, and revise filename

Revising the filename for sorting on artist, and chronological order. While author name is usually in the URL along with the art name, it's impossible to separate the two by code. Hence reliance on metadata tags to grab author (and date of painting). Would need a backup plan when it can't find author in the data - though it shouls still be named fine. Cases where author name is not in the URL, however, may cause incorrectly truncating the image name.

The embed metadata can be viewed by dropping the final image into exiftool(-k) and so forth.

kolt54321 committed:
Defaulting to zoom levels that fit in JPEG

kolt54321 committed:
Small forgotten changes

kolt54321 committed:
Requirement updates

kolt54321 commented:
Forgot to mention - autodownload.py can probably be merged into the main tile_fetch.py (which would also solve some duplication).

lovasoa commented on autodownload.py:
```diff
@@ -0,0 +1,112 @@
+import asyncio
+import tile_fetch
+import sys
+import pandas as pd
```
pandas is a huge library. It would be great if we could avoid having to add it to our dependencies

kolt54321 replied:
My friend sketched up the original code (albeit very different from now) using pandas, presumably for file handling. I'll see if I can find an alternative.

lovasoa commented on requirements.txt:
```diff
@@ -1,4 +1,9 @@
-Pillow
-aiohttp
\ No newline at end of file
+lxml~=4.5.1
+Pillow~=7.1.2
+aiohttp~=3.6.2
+pyexiv2~=2.2.0
+cssselect
+unidecode
+pandas
+html
```
It would be good if we could avoid multipying the dependencies

kolt54321 replied:
Sorry - mind to clarify? I'm not familiar with dependencies so not sure what I did wrong here.

lovasoa replied:
Can we remove unidecode and pandas and not make cssselect and html mandatory ? We can try to import them and disable the corresponding features when they are not available.

kolt54321 replied:
Should be done in latest commit... but please let me know if I missed something.

kolt54321 replied:
Any idea of how I would make cssselect conditional? I thought it was needed for scraping the metadata. As far as I know, every Arts&Culture page has metadata that can be scraped (if not for embedding, at least for final name formatting)

lovasoa commented on tile_fetch.py:
```diff
@@ -51,7 +61,8 @@ def __init__(self, url):
         url_path = urllib.parse.unquote_plus(urllib.parse.urlparse(url).path)
         self.image_slug, image_id = url_path.split('/')[-2:]
-        self.image_name = '%s - %s' % (string.capwords(self.image_slug.replace("-"," ")), image_id)
+        self.image_name = unidecode.unidecode(string.capwords(self.image_slug.replace("-"," ")))
```
Why do we need to remove non-ascii characters here ?

kolt54321 replied:
I believe I ran into some cases where the filename contained non-ascii characters, and thus download wouldn't finalize on Windows. By replacing non-ascii characters it becomes system compliant.

lovasoa replied:
What characters exactly ? All major operating systems accept unicode characters in filenames. And on the other hand, there are many ASCII characters that are rejected on windows. Can we just strip invalid characters from the name ?

kolt54321 replied:
It was a Spanish accent name, one of [these "e"s](https://www.spanishdict.com/answers/282262/how-to-type-spanish-letters-and-accents-from-paralee). I don't have the URL on hand. I would avoid stripping those letters from the name - it's part of the actual name. Replacing it with a like-letter such as "e" seems to be a better solution.

lovasoa replied:
Accented letters can be parts of filenames in all operating systems, with all but the most ancient filesystems

kolt54321 replied:
That is weird, because I think [this link](https://artsandculture.google.com/asset/cabeza-de-hierro-jaume-morera-gal%C3%ADcia/ZQGdSYXWL7WZUw) was running into issues without this piece of code to handle the characters, under the new naming scheme.

lovasoa commented on tile_fetch.py:
```diff
@@ -114,6 +125,23 @@ async def load_tiles(info, z=-1, outfile=None, quality=90):
         )
         z = len(info.tile_info) - 1
 
+    ## Ensuring image resolution fits in JPEG - two pass 
+    if info.tile_info[z].size[0] > 65535 or info.tile_info[z].size[1] > 65535:
+        print(
+            'Zoom level {r} too high for JPEG output, using next zoom level {next_z} instead'.format(
+                r=z,
+                next_z=z-1)
+        )
+        z = z-1
+        
+    if info.tile_info[z].size[0] > 65535 or info.tile_info[z].size[1] > 65535:
+        print(
+            'Zoom level {r} *still* too high for JPEG output, using next zoom level {next_z} instead'.format(
+                r=z,
+                next_z=z-1)
+        )
+        z = z-1
```
If the user has specifically requested a zoom level, we should respect it (and save as PNG if the image is too large)

But when we are choosing the zoom level automatically, we can run this instead of always choosing `z = len(info.tile_info) - 1`.

And this should be done in a loop, not with 2 successive checks.

kolt54321 replied:
We can save it as a PNG if the zoom is too large, but would then have to skip all code appending metadata to the file - PNG doesn't support it.

Correct - I'll make it into a loop.

lovasoa replied:
Yes, the code that appends metadata must be in a conditional.

kolt54321 replied:
Both of these should be done in latest commit... but please let me know if I missed something.

lovasoa commented on tile_fetch.py:
```diff
@@ -138,12 +166,35 @@ async def load_tiles(info, z=-1, outfile=None, quality=90):
+    # Taking out the author's name from the image name - authors name is appended later
+    modified_image_name = info.image_name[0:len(info.image_name)-len(author)-1]
+    
+    final_image_filename = (author + ' - ' + date + ' - ' + modified_image_name + ' - ' +info.image_id + '.jpg')
```
Save as png if the image is too large

kolt54321 replied:
Should be done in latest commit.

lovasoa commented on tile_fetch.py:
```diff
@@ -138,12 +166,35 @@ async def load_tiles(info, z=-1, outfile=None, quality=90):
+    xmp_file_obj = TaggedImage(final_image_filename)
+
+    # writes key:value one at a time, which is heavier on writes,
+    # but far more robust.
```
In what situation can tag writes fail ? Can't it be known ahead of time ?

kolt54321 replied:
My friend added this piece of catch-code - but I believe some foreign characters (Japanese, perhaps? Something unsupported on the host system) can fail. We came across one such case, but I forget what the details were behind it failing. Most times there are less than 10 tags in the page description, so I didn't feel it's a big deal.

lovasoa replied:
Can you try to find the cases that make this fail ?

Edit: ok, I just saw your comment below, thanks

kolt54321 replied:
As pere [here](https://github.com/LeoHsiao1/pyexiv2) the limit seems to be 2gb. Maybe the image before being saved is larger than that in memory?

kolt54321 replied:
> Can you try to find the cases that make this fail ?
> 
> Edit: ok, I just saw your comment below, thanks

In addition, I think [this link](https://artsandculture.google.com/asset/winter-landscape-with-bird-trap/mgG0Fs1WLFP5rg) prompted my friend to write the key:values line by line. Below is the error message I originally sent him. I think writing each value separately is better so that the whole thing doesn't fail if an XMP can't be embed, for whatever reason.

![python 2 (1)](https://user-images.githubusercontent.com/34801180/87449960-dfae5a80-c5cb-11ea-80e2-a1f73cd5d4da.png)

kolt54321 replied:
Looking at it further I am seeing a slowdown in the last piece of the code... am wondering if writing the XMP tags one by one is time consuming. Will definitely think about this...

lovasoa reviewed (changes requested):
Thanks for the PR ! I highlighted a few possible enhancements

kolt54321 commented:
One example of tags failing:

```C:\Users\h\Downloads\gapdecoder>python tile_fetch.py https://artsandculture.google.com/asset/the-kiss-gustav-klimt/HQGxUutM_F6ZGg

Downloading image meta-information...

the-kiss-gustav-klimt - zoom levels:

level  0:    480 x    481 (     1 tiles)

level  1:    959 x    962 (     4 tiles)

level  2:   1917 x   1923 (    16 tiles)

level  3:   3833 x   3846 (    64 tiles)

level  4:   7666 x   7692 (   240 tiles)

level  5:  15332 x  15384 (   930 tiles)

level  6:  30664 x  30767 (  3660 tiles)

level  7:  61328 x  61535 ( 14520 tiles)

Which level do you want to download? 

7

Downloading tiles...

100.0%

Downloaded all tiles. Saving...

Failed to add add XMP tag with key "Xmp.xmp.URL" with value "https://artsandculture.google.com/asset/the-kiss-gustav-klimt/HQGxUutM_F6ZGg"

Failed to add add XMP tag with key "Xmp.xmp.Title" with value "The Kiss"

Failed to add add XMP tag with key "Xmp.xmp.Creator" with value "Gustav Klimt"

Failed to add add XMP tag with key "Xmp.xmp.DateCreated" with value "1908-1909"

Failed to add add XMP tag with key "Xmp.xmp.PhysicalDimensions" with value "w180 x h180 cm"

Failed to add add XMP tag with key "Xmp.xmp.Type" with value "Oil on canvas"

Failed to add add XMP tag with key "Xmp.xmp.ExternalLink" with value "Explore the Belvedere online collection"

Saved the result as The Kiss Gustav Klimt - HQGxUutM_F6ZGg.jpg
```

This scenario fails to write any XMP tags, rather than a single one. I am yet not sure why.

lovasoa commented:
What is the exact RuntimeError being thrown?

kolt54321 commented:
> What is the exact RuntimeError being thrown?

Out of memory error. I'm trying to (hopefully) solve it [here](https://github.com/LeoHsiao1/pyexiv2/issues/26)

kolt54321 commented:
Small update: I've made a couple of changes (not all the suggested ones yet, but slowly) and am testing the newer version now before pushing to the branch. 

A few changes:

1. Integrate autodownload.py into tile_fetch.py. It is still slightly redundant but much less than before.
2. Saving as PNG options, and reverting to the image_name if we can't grab the author from the tags. Metadata embedding is now conditional, and though newer versions of PIL allow for embedding exif info, I'm not sure how that works.
3. Metadata embedding tries to do all at once, and if that fails, one at a time with failing tag explaining why. 
4. Zoom-fitting to JPEG is now optional (though mandatory for now when downloading from the queue with flag ```-d```), at least crudely ("choose zoom 11 to default to largest JPEG-compliant level"). We can tweak this further as we go.

So far it works well, but there's an error later one when saving the file (still trying to figure out why). Pandas is still a requirement for now though I know we should move off of it if possible.

kolt54321 commented:
https://github.com/LeoHsiao1/pyexiv2/issues/26

https://github.com/Exiv2/exiv2/issues/1248

kolt54321 committed:
Integrating autodownload into main file, PNG output, other changes

kolt54321 commented:
These new changes should solve some of the issues.

kolt54321 committed:
Update conditional

lovasoa commented:
> Not sure why the build failed.

Because unidecode was removed from requirements.txt, but is still being unconditionally imported in tile_fetch.py

kolt54321 commented:
> > Not sure why the build failed.
> 
> Because unidecode was removed from requirements.txt, but is still being unconditionally imported in tile_fetch.py

I see. I don't know how to test if there is an invalid name or non-ascii character in the name. Are you okay with me putting it back into the requirements?

kolt54321 committed:
Update requirements.txt

kolt54321 committed:
Revert "Update requirements.txt"

This reverts commit 13b86bf48f1d6c93f6e8395ac41dc9e26eb9760a.

kolt54321 committed:
Update tile_fetch.py

lovasoa commented:
As I said, I don't think there is any major platform that disallows non-ascii filenames in its default filesystem. Unicode characters in filenames are okay.

kolt54321 commented:
> As I said, I don't think there is any major platform that disallows non-ascii filenames in its default filesystem. Unicode characters in filenames are okay.

Could you try downloading with the following link? Any zoom should be fine. I took out the unidecode pieces from the code attached.

```
C:\Users\i\Downloads\gapdecoder>python tile_fetch.py --zoom 3 https://artsandculture.google.com/asset/cabeza-de-hierro-jaume-morera-gal%C3%ADcia/ZQGdSYXWL7WZUw
Downloading image meta-information...
Downloading tiles...
100.0%
Downloaded all tiles. Sing...
Traceback (most recent call last):
  File "tile_fetch.py", line 237, in <module>
    main()
  File "tile_fetch.py", line 233, in main
    loop.run_until_complete(coro)
  File "C:\Users\i\AppData\Local\Programs\Python\Python38\lib\asyncio\base_events.py", line 608, in run_until_complete
    return future.result()
  File "tile_fetch.py", line 187, in load_tiles
    xmp_file_obj = TaggedImage(final_image_filename)
  File "C:\Users\i\AppData\Local\Programs\Python\Python38\lib\site-packages\pyexiv2\core.py", line 16, in __init__
    self.img = api.open_image(filename.encode(encoding))
RuntimeError: Jaume Morera Galícia - 1891-1897 - Cabeza De Hierro Jaume Morera Gal - ZQGdSYXWL7WZUw.jpg: Failed to open the data source: No such file or directory (errno = 2)
```

lovasoa commented:
This looks like an instance of the following bug in pyexiv2: https://github.com/LeoHsiao1/pyexiv2/issues/21

kolt54321 commented:
I see. Is there a solution that doesn't involve unidecode? Since (almost) all GA&C links have metadata, and unidecode doesn't seem like a costly requirement (compared to, say, Pandas as you mentioned), I'm not averse to keeping it in the requirements.

lovasoa commented:
Don't you think more people want the filenames to be correct than having embedded meta-information, don't you think ? 
And if you still want to add the metadata, you can use the workaround mentioned in https://github.com/LeoHsiao1/pyexiv2/issues/21

kolt54321 commented:
I guess that's where we have different opinions (which is okay). I feel it is very important to know where the file came from (source URL, which is included in the metadata) as well as the author name separate from the image name. The source URL can technically be derived from the ID, which works write now appended to any other URL, but I wouldn't rely on that working moving forward. The artist name is included in the final image name, but not separate from it. The date of the painting is important to me as well, so it's no surprise I prefer to have that vs ```í``` against ```i```.

If there was a Russian or Mandarin section of GA&C, I would agree with you more. Right now it seems like GA&C is geared mostly towards english speaking users, though I could definitely be wrong about that.

I think I'm not understanding the workaround well, since using ImageData doesn't seem to work either...

kolt54321 committed:
Updated author/date created source

Instead of taking directly from the metadata, this pulls the author/date created from the page source. Cleaner and more reliable.

kolt54321 committed:
Organizes outputs into artist folders

Note - still fails where "!" is contained in the name.  Otherwise creates and updates folders with each artist's name.

kolt54321 committed:
AHK to acquire links for batch

Loops through artist pages and grabs links. Has Chrome plugin dependencies - not plug/play.

kolt54321 committed:
Updated requirements

kolt54321 commented:
Unfortunately, dependencies are still sloppy... my non-comp sci background showing. I haven't had a chance to replace Pandas - I know you preferred an alternative. For now this works (though not ideal) for batch downloading, and includes code to "skip" (aka ignore) any errors as videos and some other GAC pages will throw an error and stop the code. Included are some scripts to both gather links to begin with, as well as organize the outputs into folders for each artist (outputs organized by %artist% - %date% - %name% - %id%).

kolt54321 committed:
Push Regex Change
